### PR TITLE
Closes AgileVentures/MetPlus_tracker#399

### PR DESCRIPTION
**Pull request by jeanwei**
- add index to job_applications table for uniqueness at database level
- add validation to job_application model for uniqueness at application level
- update jobs_controller to rescue the above exceptions with flash[:alert] & redirect

### CODE REVIEW

File: app/controllers/jobs_controller.rb [hunk 1]
```diff
@@ -167,24 +167,28 @@ def apply
 			return
 		end
 
-		if pets_user == @job_seeker
-			begin
-				@job.apply @job_seeker
-				Event.create(:JS_APPLY, @job.last_application_by_job_seeker(@job_seeker))
-			rescue Exception => e
-				flash[:alert] = "Unable to apply at this moment, please try again."
+		if pets_user == @job_seeker || @job_seeker.job_developer  # to be removed once authorize is set properly
```
Comment (patmbolger) [inline app/controllers/jobs_controller.rb]: I think this means to be:
```
if pets_user == @job_seeker || pets_user == @job_seeker.job_developer
```

File: app/controllers/jobs_controller.rb [hunk 2]
```diff
@@ -167,24 +167,28 @@ def apply
...
+			begin 
+				job_app = @job.apply @job_seeker
+			# ActiveRecord::RecordInvalid is raised when validation at model level fails
+			# ActiveRecord::RecordNotUnique is raised when unique index constraint on the database is violated
+			rescue ActiveRecord::RecordInvalid, ActiveRecord::RecordNotUnique  => e
+				flash[:alert] = "Invalid action due to duplicated application"
```
Comment (patmbolger) [inline app/controllers/jobs_controller.rb]: Suggest changing to:
```
flash[:alert] = "#{@job_seeker.full_name(last_name_first: false) has already applied to this job."
```

File: spec/controllers/jobs_controller_spec.rb [hunk 3]
```diff
@@ -839,6 +831,22 @@
...
+        get :apply, :job_id => @job.id, :user_id => job_seeker.id
+      end
+      it "shows flash[:alert]" do
+        expect(flash[:alert]).to be_present.and eq "Invalid action due to duplicated application"
+      end
+      it "redirects to job" do
+        expect(response).to redirect_to(:action => 'show', :id => @job.id)
+      end
+    end
     describe 'invalid application as job developer' do
```
Comment (patmbolger) [inline spec/controllers/jobs_controller_spec.rb]: in the next section ('invalid application as job developer'), I think the `before` block should be:
```
before :each do
        agency = FactoryGirl.create(:agency)
        job_developer = FactoryGirl.create(:job_developer, agency: agency)
        job_seeker.assign_job_developer(job_developer, agency)
        job_developer2 = FactoryGirl.create(:job_developer, agency: agency)
        allow(controller).to receive(:current_user).and_return(job_developer2)
        get :apply, :job_id => @job.id, :user_id => job_seeker.id
      end
```
This is to test the logic in line 170 in jobs_controller.

**Comment by patmbolger** [comment]
PR reviewed - please see comments.

**Comment by joaopapereira** [comment]
Generally looks good, please rebase and address comments made by Patrick

**Comment by jeanwei** [comment]
Thanks for the review. Code has updated accordingly. Please review

**Comment by patmbolger** [comment]
Changes look good - PR looks OK to merge.

**Comment by jeanwei** [comment]
There is a conflict. i am having problem to resolve it

**Comment by jeanwei** [comment]
screen shot 2016-09-10 at 10 28 15
The hist in my local is different from the pr.. how to resolve it?

**Comment by jeanwei** [comment]
Oh no.. suddenly arun get pulled into this review for resolving the conflicts .. =.=

**Comment by joaopapereira** [comment]
@jeanwei do you believe you can create a new PR only with your change?

**Comment by jeanwei** [comment]
@joaopapereira meaning I need to create a new PR? What is the better way to close it?

**Pull request closed without merging**